### ISSUE TITLE
Copying tensors to GPU fails non-deterministically

### ISSUE THREAD

**Issue by kriskorrel-cw** [tensorflow/tensorflow#37900]
**System information** 
- Have I written custom code (as opposed to using a stock
example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g.,
Linux Ubuntu 16.04): Ubuntu 16.04.6 LTS (GNU/Linux 4.4.0-174-generic x86_64)
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if
the issue happens on mobile device: No
- TensorFlow installed from (source or
binary): `pip install tensorflow==2.2.0rc1`
- TensorFlow version (use command below): v2.2.0-rc0-43-gacf4951a2f 2.2.0-rc1
- Python version: Python 3.7.3
- CUDA/cuDNN version: 10.1, V10.1.243
- GPU model and memory: 2x GeForce GTX 1080 8GB 

**Describe the current behavior**
I have a system with two GPUs. Since I need the fix of https://github.com/tensorflow/tensorflow/issues/33929 I have upgraded from 2.1 to 2.2.0rc0 and 2.2.0rc1. In both cases, I sometimes get the following exception when trying to train on a GPU:
```
Traceback (most recent call last):
  File "/data/personal/username/deployed/project/project/bin/train.py", line 86, in <module>
    main()
  File "/data/personal/username/deployed/project/project/bin/train.py", line 82, in main
    train()  # pragma: no cover
  File "/data/personal/username/deployed/project/project/bin/train.py", line 75, in train
    callbacks=callbacks,
  File "/data/personal/username/deployed/project/project/objects/models/KerasModel.py", line 190, in train_on_generator
    callbacks=callbacks,
  File "/home/username/.local/share/virtualenvs/project-1BQtdZDZ/lib/python3.7/site-packages/tensorflow/python/keras/engine/training.py", line 65, in _method_wrapper
    return method(self, *args, **kwargs)
  File "/home/username/.local/share/virtualenvs/project-1BQtdZDZ/lib/python3.7/site-packages/tensorflow/python/keras/engine/training.py", line 783, in fit
    tmp_logs = train_function(iterator)
  File "/home/username/.local/share/virtualenvs/project-1BQtdZDZ/lib/python3.7/site-packages/tensorflow/python/eager/def_function.py", line 580, in __call__
    result = self._call(*args, **kwds)
  File "/home/username/.local/share/virtualenvs/project-1BQtdZDZ/lib/python3.7/site-packages/tensorflow/python/eager/def_function.py", line 644, in _call
    return self._stateless_fn(*args, **kwds)
  File "/home/username/.local/share/virtualenvs/project-1BQtdZDZ/lib/python3.7/site-packages/tensorflow/python/eager/function.py", line 2420, in __call__
    return graph_function._filtered_call(args, kwargs)  # pylint: disable=protected-access
  File "/home/username/.local/share/virtualenvs/project-1BQtdZDZ/lib/python3.7/site-packages/tensorflow/python/eager/function.py", line 1665, in _filtered_call
    self.captured_inputs)
  File "/home/username/.local/share/virtualenvs/project-1BQtdZDZ/lib/python3.7/site-packages/tensorflow/python/eager/function.py", line 1746, in _call_flat
    ctx, args, cancellation_manager=cancellation_manager))
  File "/home/username/.local/share/virtualenvs/project-1BQtdZDZ/lib/python3.7/site-packages/tensorflow/python/eager/function.py", line 598, in call
    ctx=ctx)
  File "/home/username/.local/share/virtualenvs/project-1BQtdZDZ/lib/python3.7/site-packages/tensorflow/python/eager/execute.py", line 60, in quick_execute
    inputs, attrs, num_outputs)
tensorflow.python.framework.errors_impl.UnknownError: 2 root error(s) found.
  (0) Unknown:  InternalError: Failed copying input tensor from /job:localhost/replica:0/task:0/device:CPU:0 to /job:localhost/replica:0/task:0/device:GPU:0 in order to run AddV2: Attempted to set tensor for existing mirror. [Op:AddV2]
Traceback (most recent call last):

  File "/home/username/.local/share/virtualenvs/project-1BQtdZDZ/lib/python3.7/site-packages/tensorflow/python/ops/script_ops.py", line 241, in __call__
    return func(device, token, args)

  File "/home/username/.local/share/virtualenvs/project-1BQtdZDZ/lib/python3.7/site-packages/tensorflow/python/ops/script_ops.py", line 130, in __call__
    ret = self._func(*args)

  File "/home/username/.local/share/virtualenvs/project-1BQtdZDZ/lib/python3.7/site-packages/tensorflow/python/autograph/impl/api.py", line 309, in wrapper
    return func(*args, **kwargs)

  File "/data/personal/username/deployed/project/project/objects/generators/GeneratorRetinaNet.py", line 85, in _getitem_pre_anchors
    image, node = self.get_prepped_node(idx=idx)

  File "/data/personal/username/deployed/project/project/objects/generators/Generator.py", line 252, in get_prepped_node
    node = self.get_node(idx=idx)

  File "/data/personal/username/deployed/project/project/objects/generators/GeneratorObjectDetection.py", line 97, in get_node
    node = super(GeneratorObjectDetection, self).get_node(idx=idx)

  File "/data/personal/username/deployed/project/project/objects/generators/Generator.py", line 376, in get_node
    return self._dataFetcher.get_node(idx)

  File "/data/personal/username/deployed/project/project/objects/generators/DataFetchers/DataFetcher.py", line 117, in get_node
    nodes = self.get_nodes(node_index, node_index + 1)

  File "/home/username/.local/share/virtualenvs/project-1BQtdZDZ/lib/python3.7/site-packages/tensorflow/python/ops/math_ops.py", line 997, in binary_op_wrapper
    return func(x, y, name=name)

  File "/home/username/.local/share/virtualenvs/project-1BQtdZDZ/lib/python3.7/site-packages/tensorflow/python/ops/math_ops.py", line 1276, in _add_dispatch
    return gen_math_ops.add_v2(x, y, name=name)

  File "/home/username/.local/share/virtualenvs/project-1BQtdZDZ/lib/python3.7/site-packages/tensorflow/python/ops/gen_math_ops.py", line 480, in add_v2
    _ops.raise_from_not_ok_status(e, name)

  File "/home/username/.local/share/virtualenvs/project-1BQtdZDZ/lib/python3.7/site-packages/tensorflow/python/framework/ops.py", line 6653, in raise_from_not_ok_status
    six.raise_from(core._status_to_exception(e.code, message), None)

  File "<string>", line 3, in raise_from

tensorflow.python.framework.errors_impl.InternalError: Failed copying input tensor from /job:localhost/replica:0/task:0/device:CPU:0 to /job:localhost/replica:0/task:0/device:GPU:0 in order to run AddV2: Attempted to set tensor for existing mirror. [Op:AddV2]
```

The odd thing about this is, that is is not deterministic behavior. At one moment it may not work. When I kill the process, wait a minute and try again, it sometimes works. 
In general, retrying one or two times will fix the issue, but currently I'm only successful to train about 10% of the time. Is this a known issue?

Note that in (TF1.13,) TF2.0 and TF2.1, this seemed to work fine for me.

Before training, I always use `os.environ['CUDA_VISIBLE_DEVICES'] = '0` or `'1'`, depending on the GPU I want to use. I have tried replacing this with `tf.config.set_visible_devices` to no avail.

For completeness, I do not use TPUs or distributed strategies.

**Comment by goldiegadde** [comment]
@kriskorrel-cw Can you please add minimum code snippet preferrably in colab ?

**Comment by kriskorrel-cw** [comment]
As I have written (a lot of) custom code, I currently cannot reproduce the issue with a small code snippet.
I have tried with one of the tensorflow tutorials, in which case the problem has not occurred (yet).

**Comment by kriskorrel-cw** [comment]
Alright, here are my not-so-scientific findings:
1. The problem appears to be code-specific, which is demonstrated by the fact that running some tensorflow tutorial code seemed to work fine so far.
2. The problem is very non-deterministic. Yesterday I have tried to run the exact same training script multiple times, each time making sure the the program is fully killed and the GPU is freed before I try again. I had to retry for 9 times before I succeeded in running the script. After that, it failed again for 6 times, after which it succeeded again.
3. I am not sure how machine/environment dependent it is. We have two (almost) identical machines, both with two GPUs. On one machine, I can hardly run the train script. On the other, it works about 90% of the time (very rough estimate). 
4. Luckily for me, the problems seems to disappear when upgrading from `v2.2.0-rc0-43-gacf4951a2f 2.2.0-rc1` to tf-nightly: `v1.12.1-28047-g983a5e245e 2.2.0-dev20200325`. This might also help in pinpointing the problem.

**Comment by gadagashwini** [comment]
@kriskorrel-cw, Can we close since the problem is disappeared in tf-nightly.

**Comment by kriskorrel-cw** [comment]
> @kriskorrel-cw, Can we close since the problem is disappeared in tf-nightly.

Would be fine by me

**Comment by goldiegadde** [comment]
Hi @kriskorrel-cw , thanks for reporting the issue, and also verifying with tf-nightly. At this point without a code snippet it is hard to pinpoint the issue. 
Please feel free to reopen if the issue reappears or if you have a reproable code snippet.

**Comment by kriskorrel-cw** [comment]
I am reopening this is as the issue persists to this day.
Since tensorflow 2.2.0 is now out, I have tried installing this version and retrying whether I can train a model (consistently), and it appears not. However, for the past few weeks, and still today, I can train models perfectly fine with any version of `tf-nightly`. Are there somehow any differences between the two packages?

I also had another look at the stack trace which I posted. I find the following part very odd:
```
  File "/data/personal/username/deployed/project/project/objects/generators/GeneratorRetinaNet.py", line 85, in _getitem_pre_anchors
    image, node = self.get_prepped_node(idx=idx)

  File "/data/personal/username/deployed/project/project/objects/generators/Generator.py", line 252, in get_prepped_node
    node = self.get_node(idx=idx)

  File "/data/personal/username/deployed/project/project/objects/generators/GeneratorObjectDetection.py", line 97, in get_node
    node = super(GeneratorObjectDetection, self).get_node(idx=idx)

  File "/data/personal/username/deployed/project/project/objects/generators/Generator.py", line 376, in get_node
    return self._dataFetcher.get_node(idx)

  File "/data/personal/username/deployed/project/project/objects/generators/DataFetchers/DataFetcher.py", line 117, in get_node
    nodes = self.get_nodes(node_index, node_index + 1)
```
This is all custom written code.

For some background: We use a `tf.data.Dataset` for generating data and then train using `model.fit(x=my_dataset)`.
Because of reasons, basically all our data operations/preparations need to be performed with python code, which cannot be easily converted to tensorflow operations nor automatically converted using autograph.
Our tf.data.Dataset thermore basically consists in a map, which executes a `tf.py_function`. 
The part of the stack trace above are all methods that are be called within this py_function. I don't see, however, how this has anything to do with the GPU. As far as I know, no tf.data.Dataset operations should have anything to do with the GPU.

**Comment by kriskorrel-cw** [comment]
Update: as can be seen in the stack trace, `nodes = self.get_nodes(node_index, node_index + 1)` is the last line of custom code that appears. For debugging, I have replaced this line with
```python
print(node_index, type(node_index), node_index.device)
test = node_index + 1
nodes = self.get_nodes(node_index, node_index + 1)
```
Which outputs `tf.Tensor(42584, shape=(), dtype=int64) <class 'tensorflow.python.framework.ops.EagerTensor'> /job:localhost/replica:0/task:0/device:CPU:0` after which a similar stack trace will appear as before, but indeed crashing on the line `test = node_index + 1` instead.

Note that `node_index` is the input to the py_function and is generated by `tf.data.Dataset.range(dataset_length)`. 

If I add a line `node_index = node_index.numpy()`, my code now seems to work fine, but I'm not sure whether the issue is "solved" with this.
I still have a lot of questions. Why does it work without this statement on tf-nightly, but not on tensorflow 2.2.0 (rc0/rc1/rc2)? Why does this work when I train on a CPU and not when training on a GPU? Why is this tensor being copied to GPU? Or at least I assume it is, based on the stack trace.

Note that the output of the print statement is the same for both tensorflow versions, and both when I'm training on a CPU or GPU.

**Comment by mberezo** [comment]
Getting the same error today.  Going to try tf-nightly for now.  Want to collab @kriskorrel-cw ?

**Comment by kriskorrel-cw** [comment]
@mberezo I would be interested to hear whether your setup and cause of the issue is similar to mine. Maybe this could help in writing a minimal reproducing code snippet. I was unable to do this before.

**Comment by amahendrakar** [comment]
@kriskorrel-cw, @mberezo,
Could you please provide a minimal sample code to reproduce the issue reported here. Thanks!

**Comment by kriskorrel-cw** [comment]
> @kriskorrel-cw, @mberezo,
> Could you please provide a minimal sample code to reproduce the issue reported here. Thanks!

Currently, I'm not able to do this. I cannot share my own code, and I have tried twice to replicate the issue with minimal code, but was unable to do so. When I have more time, I will try again with my own codebase, and strip it down as much as possible to get a minimal reproducing snippet.

I was hoping that you could figure out, or point in a direction based on the stack trace, the situation I described, or the fact the the problem is not present in tf-nightly, but it is in tensorflow.

**Comment by AroMorin** [comment]
I am experiencing this issue as well. I am training with a tf.dataset and using dataset.map() + a tf.py_function() to process the dataset elements. Can't share code. Maybe I'll try tf-nightly.

**Comment by AroMorin** [comment]
Update: If I put this line inside a `with tf.device('/device:cpu:0'):` block, then I don't get the error and everything works well. The idea was to prevent a copy from CPU to GPU, and this achieves that idea. Not sure what happened, but it's a quick-fix.

**Comment by AroMorin** [comment]
Update 2: Another part of my pipeline reproduced this issue. Instead of putting it in a with `tf.device()` block as above, I turned OFF "num_parallel_calls" in my dataset.map() function. The error went away. It seems there is something about the `num_parallel_calls` attribute that triggers this error.

**Comment by kriskorrel-cw** [comment]
@AroMorin Thanks for sharing. Good to know that `num_parallel_calls` is also essential for producing this bug. Unfortunately, with that, I'm still not able to reproduce. [This](https://gist.github.com/kriskorrel-cw/72317d7007ee3d3688f884b7a61a4257) is what I tried, btw.

Is it for you also true that `img = img.numpy()` will 'fix' the issue?

**Comment by ymodak** [comment]
@kriskorrel-cw The example works with TF 2.2 [gist](https://colab.research.google.com/gist/ymodak/8f56b5df2f4cf1f018a711ad3d3b832a/untitled21.ipynb#scrollTo=SK9SMbuWmrRs)
Please update the issue when you get a chance and error is reproduced.
Thank you.

@AroMorin  If you can share your example code to reproduce this behavior we can take a look?
Thanks!

**Comment by kriskorrel-cw** [comment]
@ymodak earlier I said 
> @AroMorin Thanks for sharing. Good to know that num_parallel_calls is also essential for producing this bug. Your setup seems to be similar to mine. Unfortunately, with that, I'm still not able to reproduce. This is what I tried, btw.

Here, I linked to a gist which would work, i.e., no exception could be reproduced with this gist. However, I tested this by running on Google Colab. This, you have also done yourself and found out that there was no problem with it. Today, however, I had the chance to test this code snippet on my own machine, and here it would throw the same exception! So the problem is probably machine/environement-dependent as I already suggested earlier. If you need to know anything about my system, please let me know. Some information is already in the first comment of this issue.

So, to reiterate, when running [this exact script](https://gist.github.com/kriskorrel-cw/72317d7007ee3d3688f884b7a61a4257) on my machine, using tf `v2.2.0-rc4-8-g2b96f3662b 2.2.0` I (stochastically) get the following (complete output)[https://gist.github.com/kriskorrel-cw/c60035e8f2bcc56337edfd535648f4e6]

**Comment by sanjoy** [comment]
Hi @kriskorrel-cw,

Unfortunately, I don't think we'll be able to productively debug a machine dependent issue that we can't reproduce.

**Comment by sanjoy** [comment]
Hi @AroMorin,

Were you able to reproduce the issue using tf-nightly?  If yes, do you mind sharing a reproducer?

**Comment by kriskorrel-cw** [comment]
> Hi @kriskorrel-cw,
> 
> Unfortunately, I don't think we'll be able to productively debug a machine dependent issue that we can't reproduce.

I understand that. At least there seems to be a workaround (calling `.numpy()` within the py_function) to make it work on tf 2.2.0. Hopefully someone or something will allow to debug this in the future.

**Comment by pshved** [comment]
I am also able to reproduce the error with my code in TF 2.2. 
I looked up the place in the TF code with the error message, and found this commit to 2.3 that sounds like it might be relevant to the problem https://github.com/tensorflow/tensorflow/commit/6aee08c765b08f1f91663a32e89913336def2ef2

So I upgraded to tf-nightly, and I'm no longer seeing the issue.

@sanjoy when many users report a flaky issue with the same error message, there are things one can do without the repro.  One way could be to find the line with the error message in the code search, and try to ask, "what are the potential race conditions that would trigger this error message?" This approach worked for me pretty often.

**Comment by xnio94** [comment]
I encounter this error in a different context, it was hard to reproduce at first but after some work, i was able to make a [minimum repro code snippet ](https://colab.research.google.com/drive/1RErUzZXUUfg-VNjjczspBex17R2KSfrl?usp=sharing) 
(make sure to use a GPU to reproduce the error)
hope it helps
@sanjoy

**Comment by xnio94** [comment]
other remarques :

- when i re-run the script after the error rise, it will work fine

- when i encountered the error in my project it was non-deterministic. but after making the script above the error seems to accrue all of the time.

**Comment by sanjoy** [comment]
@xnio94 thanks for the reproducer, but it seems to work fine on my machine.

@jaingaurav any guesses on why we'd get this error: "AddV2: Attempted to set tensor for existing mirror. [Op:AddV2]"?

**Comment by jaingaurav** [comment]
@sanjoy: I think the description in 6aee08c that @pshved noted explains it correctly. There are basically some race conditions that we need to more gracefully handle. From what I'm reading the issue is not reproducible after that fix correct?

**Comment by pshved** [comment]
Yep @jaingaurav I have never seen this issue after I upgraded to the nightly version of TF, which contained 6aee08c.  I see it sometimes when I forget to switch to the nightly version and when my python env still remains at TF 2.2.  Which seems to indicate that 6aee08c does fix this problem.

**Comment by arainboldt** [comment]
FWIW, I'm getting this issue with the below code:

```
import tensorflow as tf
import numpy as np
import multiprocessing as mp

test = np.random.binomial(1,.3,80000).reshape((1000,80))

def run_release_gpu(func):
    def parallel_wrapper(output_dict, *argv, **kwargs):
        ret = func(*argv, **kwargs)
        # print(ret)
        if ret is not None:
            output_dict['ret'] = ret

    def outer_wrapper(*argv, **kwargs):
        same_process = kwargs.pop('same_process', False)
        if same_process:
            return func(*argv, **kwargs)

        with mp.Manager() as manager:
            output = manager.dict()
            args = (output,) + argv
            p = mp.Process(target=parallel_wrapper, args=args, kwargs=kwargs)
            p.start()
            p.join()
            ret_val = output.get('ret', None)

        return ret_val

    return outer_wrapper


@run_release_gpu
def tf_hamming(a, b):
    l = a.shape[1]
    res = tf.math.divide_no_nan(
        tf.reduce_sum(
            tf.cast(
                tf.not_equal(tf.expand_dims(a, axis=1), tf.expand_dims(b, axis=0)),
                tf.float16)
            , axis=-1),
        l)
    return res
```

I have the same GPU as OP. 
python 3.8.3
Ubuntu 20.04 
Tensorflow 2.3.1

```
File "/home/.../.local/share/virtualenvs/tf/lib/python3.8/site-packages/tensorflow/python/util/dispatch.py", line 201, in wrapper
    return target(*args, **kwargs)
  File "/home/.../.local/share/virtualenvs/tf/lib/python3.8/site-packages/tensorflow/python/ops/array_ops.py", line 435, in expand_dims_v2
    return gen_array_ops.expand_dims(input, axis, name)
  File "/home/.../.local/share/virtualenvs/tf/lib/python3.8/site-packages/tensorflow/python/ops/gen_array_ops.py", line 2313, in expand_dims
    _ops.raise_from_not_ok_status(e, name)
  File "/home/.../.local/share/virtualenvs/tf/lib/python3.8/site-packages/tensorflow/python/framework/ops.py", line 6843, in raise_from_not_ok_status
    six.raise_from(core._status_to_exception(e.code, message), None)
  File "<string>", line 3, in raise_from
tensorflow.python.framework.errors_impl.InternalError: Failed copying input tensor from /job:localhost/replica:0/task:0/device:CPU:0 to /job:localhost/replica:0/task:0/device:GPU:0 in order to run ExpandDims: Dst tensor is not initialized. [Op:ExpandDims]
```

**Comment by ydennisy** [comment]
I have also come across this on TF2.4, I had to keep reducing down the size of various params (vocabs, embedding dims) until the model ran.

**Comment by jvishnuvardhan** [comment]
@kriskorrel-cw Based on the comments above, this should have been resolved by `TF2.4.1` or above. Can you please check and let us know whether this is resolved or not.

@xnio94 If this is still an issue for you, can you share the data to test your standalone code? Thanks! 

Please feel free to close the issue If this was resolved with recent `TF` versions. Thanks!

**Comment by kriskorrel-cw** [comment]
> @kriskorrel-cw Based on the comments above, this should have been resolved by `TF2.4.1` or above. Can you please check and let us know whether this is resolved or not.
> 
> @xnio94 If this is still an issue for you, can you share the data to test your standalone code? Thanks!
> 
> Please feel free to close the issue If this was resolved with recent `TF` versions. Thanks!

"unfortunately" I currently can't reproduce the issue with TF 2.2.0. I don't know the reason of this.
However, if the issue seems to be pinpointed and fixed in TF 2.4.1, I assume that this is correct, and wil now close the issue.

**Comment by guba-odudkin** [comment]
Having the same issue on TF 2.6.0. Using Laptop RTX 3000
